capsules/extra/src/i2c_nmea.rs

```diff
@@ -0,0 +1,175 @@
+                });
+                return;
+            }
+        };
+        let mut nmea_offset = self.nmea_offset.get();
+
+        if let Some(location) = string.find('$') {
+            // This is the start of a new sentence
+
+            if (nmea_offset + location) > nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
+                self.nmea_offset.set(0);
+                self.sentence_buffer.replace(nmea_buf);
+
+                if let Err((e, buf)) = self.i2c.read(buffer, i2c_buf_len) {
+                    self.i2c_buffer.replace(buf);
+
+                    self.client.map(|call| {
+                        call.callback(self.sentence_buffer.take().unwrap(), 0, Err(e.into()));
+                    });
+                }
+                return;
+            }
+
+            // First copy the remainder of the sentence to the buffer
+            nmea_buf[nmea_offset..(nmea_offset + location)].copy_from_slice(&buffer[0..location]);
+            nmea_offset += location;
+
+            // Now parse the sentence
+            let sentence = match str::from_utf8(&nmea_buf[0..nmea_offset]) {
+                Ok(utf8) => utf8,
+                Err(_e) => {
+                    self.i2c_buffer.replace(buffer);
+
+                    self.client.map(|call| {
+                        call.callback(nmea_buf, 0, Err(ErrorCode::NOSUPPORT));
+                    });
+                    return;
+                }
+            };
+
+            if sentence.starts_with('$') {
+                // At this point we have a sentence with a `$` at the start.
+                // We report it back to the caller.
+                // We loose the rest of the data we just read though
+                self.i2c_buffer.replace(buffer);
+                self.nmea_offset.set(0);
+
+                self.client.map(|call| {
+                    call.callback(nmea_buf, nmea_offset, Ok(()));
+                });
+
+                return;
+            } else {
+                // The sentence didn't start with `$`. This usually occurs
+                // if we start reading mid-sentence. So we just try again and
+                // get the next one.
+                nmea_offset = 0;
+
+                // Copy the rest of the data
+                let size = i2c_buf_len - location;
+                nmea_buf[nmea_offset..(nmea_offset + size)].copy_from_slice(&buffer[location..]);
+                nmea_offset += size;
+            }
+        } else {
+            if (nmea_offset + i2c_buf_len) > nmea_buf.len() {
+                // We will overflow our buffer, just drop the packet and try again
+                nmea_offset = 0;
+            }
+
+            // This is not a new sentence, copy the entire string to our buffer
+            nmea_buf[nmea_offset..(nmea_offset + i2c_buf_len)].copy_from_slice(buffer);
+            nmea_offset += i2c_buf_len;
```

Comment:
I don't understand why there would always be `i2c_buf_len` valid bytes in the buffer.

Reply:
We will only get the callback when we finish reading all of the bytes we requested (or if there was an error)

Reply:
So the expectation is the NMEA device stretches the i2c clock? What happens if the the total message is, say, 25 bytes, so the first 24 are read, then the second read gets the 25th byte, but is still waiting on 23 other bytes?

Reply:
A NMEA device will continuously send sentences.

So if the message is 25 bytes, we read 24 bytes. Then we read another 24 bytes. That provides a sentence and some part of the next sentence.

Reply:
NMEA 0183 is designed to work over serial. So it's just a lot of serial prints